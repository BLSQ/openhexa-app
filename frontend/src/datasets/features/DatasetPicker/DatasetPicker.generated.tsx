import * as Types from '../../../graphql/types';

import { gql } from '@apollo/client';
export type DatasetPicker_WorkspaceFragment = { __typename?: 'Workspace', datasets: { __typename?: 'DatasetLinkPage', items: Array<{ __typename?: 'DatasetLink', id: string, dataset: { __typename?: 'Dataset', slug: string, name: string } }> } };

export const DatasetPicker_WorkspaceFragmentDoc = gql`
    fragment DatasetPicker_workspace on Workspace {
  datasets {
    items {
      id
      dataset {
        slug
        name
      }
    }
  }
}
    `;