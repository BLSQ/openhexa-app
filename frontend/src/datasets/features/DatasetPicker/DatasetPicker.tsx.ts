import { graphql } from "graphql/gql";

export const DatasetPickerWorkspaceDoc = graphql(`
fragment DatasetPicker_workspace on Workspace {
  datasets {
    items {
      id
      dataset {
        slug
        name
      }
    }
  }
}
`);
