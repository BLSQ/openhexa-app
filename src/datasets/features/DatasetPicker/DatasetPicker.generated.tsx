import * as Types from '../../../graphql-types';

import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
const defaultOptions = {} as const;
export type DatasetPickerQueryVariables = Types.Exact<{
  slug: Types.Scalars['String']['input'];
}>;


export type DatasetPickerQuery = { __typename?: 'Query', workspace?: { __typename?: 'Workspace', slug: string, datasets: { __typename?: 'DatasetLinkPage', items: Array<{ __typename?: 'DatasetLink', id: string, dataset: { __typename?: 'Dataset', slug: string, name: string } }> } } | null };

export type DatasetPicker_WorkspaceFragment = { __typename?: 'Workspace', datasets: { __typename?: 'DatasetLinkPage', items: Array<{ __typename?: 'DatasetLink', id: string, dataset: { __typename?: 'Dataset', slug: string, name: string } }> } };

export const DatasetPicker_WorkspaceFragmentDoc = gql`
    fragment DatasetPicker_workspace on Workspace {
  datasets {
    items {
      id
      dataset {
        slug
        name
      }
    }
  }
}
    `;
export const DatasetPickerDocument = gql`
    query DatasetPicker($slug: String!) {
  workspace(slug: $slug) {
    slug
    ...DatasetPicker_workspace
  }
}
    ${DatasetPicker_WorkspaceFragmentDoc}`;

/**
 * __useDatasetPickerQuery__
 *
 * To run a query within a React component, call `useDatasetPickerQuery` and pass it any options that fit your needs.
 * When your component renders, `useDatasetPickerQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useDatasetPickerQuery({
 *   variables: {
 *      slug: // value for 'slug'
 *   },
 * });
 */
export function useDatasetPickerQuery(baseOptions: Apollo.QueryHookOptions<DatasetPickerQuery, DatasetPickerQueryVariables> & ({ variables: DatasetPickerQueryVariables; skip?: boolean; } | { skip: boolean; }) ) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useQuery<DatasetPickerQuery, DatasetPickerQueryVariables>(DatasetPickerDocument, options);
      }
export function useDatasetPickerLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<DatasetPickerQuery, DatasetPickerQueryVariables>) {
          const options = {...defaultOptions, ...baseOptions}
          return Apollo.useLazyQuery<DatasetPickerQuery, DatasetPickerQueryVariables>(DatasetPickerDocument, options);
        }
export function useDatasetPickerSuspenseQuery(baseOptions?: Apollo.SuspenseQueryHookOptions<DatasetPickerQuery, DatasetPickerQueryVariables>) {
          const options = {...defaultOptions, ...baseOptions}
          return Apollo.useSuspenseQuery<DatasetPickerQuery, DatasetPickerQueryVariables>(DatasetPickerDocument, options);
        }
export type DatasetPickerQueryHookResult = ReturnType<typeof useDatasetPickerQuery>;
export type DatasetPickerLazyQueryHookResult = ReturnType<typeof useDatasetPickerLazyQuery>;
export type DatasetPickerSuspenseQueryHookResult = ReturnType<typeof useDatasetPickerSuspenseQuery>;
export type DatasetPickerQueryResult = Apollo.QueryResult<DatasetPickerQuery, DatasetPickerQueryVariables>;