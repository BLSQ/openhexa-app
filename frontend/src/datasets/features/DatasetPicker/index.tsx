import DatasetPicker from "./DatasetPicker";

export default DatasetPicker;
