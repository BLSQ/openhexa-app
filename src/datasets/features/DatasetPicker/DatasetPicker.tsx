import { gql, useQuery } from "@apollo/client";
import { useCallback, useMemo, useState } from "react";
import { useTranslation } from "next-i18next";
import useDebounce from "core/hooks/useDebounce";
import { Combobox } from "core/components/forms/Combobox";
import { DatasetPickerQuery } from "./DatasetPicker.generated";

type Option = {
  id: string;
  dataset: { slug: string; name: string };
};

type DatasetPickerProps = {
  value?: string;
  workspaceSlug: string;
  placeholder?: string;
  onChange(value?: Option): void;
  required?: boolean;
  disabled?: boolean;
  withPortal?: boolean;
};

const DatasetPicker = (props: DatasetPickerProps) => {
  const { t } = useTranslation();
  const {
    workspaceSlug,
    value,
    disabled = false,
    required = false,
    withPortal = false,
    onChange,
    placeholder = t("Select dataset"),
  } = props;

  const { data, loading } = useQuery<DatasetPickerQuery>(
    gql`
      query DatasetPicker($slug: String!) {
        workspace(slug: $slug) {
          slug
          ...DatasetPicker_workspace
        }
      }
      ${DatasetPicker.fragments.workspace}
    `,
    { variables: { slug: workspaceSlug } },
  );

  const [query, setQuery] = useState("");
  const debouncedQuery = useDebounce(query, 150);

  const options = useMemo(() => {
    const lowercaseQuery = debouncedQuery.toLowerCase();

    return (
      data?.workspace?.datasets.items.filter((item) => {
        return item.dataset.name.toLowerCase().includes(lowercaseQuery);
      }) ?? []
    );
  }, [data, debouncedQuery]);

  const displayValue = useCallback(
    (option: Option) => (option ? option.dataset.name : ""),
    [],
  );

  const comboBoxValue = useMemo(() => {
    return options.find((option) => option.dataset.slug === value) ?? null;
  }, [value, options]);

  return (
    <Combobox
      required={required}
      onChange={onChange}
      loading={loading}
      withPortal={withPortal}
      displayValue={displayValue}
      by={(a, b) => a && b && a.dataset.slug === b.dataset.slug}
      onInputChange={useCallback(
        (event: any) => setQuery(event.target.value),
        [],
      )}
      value={comboBoxValue as any}
      placeholder={placeholder}
      onClose={useCallback(() => setQuery(""), [])}
      disabled={disabled}
    >
      {options.map((option: Option) => (
        <Combobox.CheckOption key={option.id} value={option}>
          {option.dataset.name}
        </Combobox.CheckOption>
      ))}
    </Combobox>
  );
};

DatasetPicker.fragments = {
  workspace: gql`
    fragment DatasetPicker_workspace on Workspace {
      datasets {
        items {
          id
          dataset {
            slug
            name
          }
        }
      }
    }
  `,
};

export default DatasetPicker;
