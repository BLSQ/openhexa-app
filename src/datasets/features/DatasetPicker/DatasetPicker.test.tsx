import { useQuery } from "@apollo/client";
import { faker } from "@faker-js/faker";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import DatasetPicker from "./DatasetPicker";

jest.mock("@apollo/client", () => ({
  __esModule: true,
  useQuery: jest.fn(),
  gql: jest.fn(() => "GQL"),
}));

const useQueryMock = useQuery as jest.Mock;

const WORKSPACE = {
  slug: faker.string.uuid(),
  datasets: {
    items: [
      {
        id: faker.string.uuid(),
        dataset: {
          slug: faker.lorem.slug(5),
          name: "Dataset 1",
        },
      },
      {
        id: faker.string.uuid(),
        dataset: {
          slug: faker.lorem.slug(5),
          name: "Dataset 2",
        },
      },
    ],
  },
};

describe("DatasetPicker", () => {
  it("display all dataset", async () => {
    const user = userEvent.setup();
    useQueryMock.mockReturnValue({
      loading: true,
      data: {
        workspace: WORKSPACE,
      },
    });
    const onChange = jest.fn();

    const { container } = render(
      <DatasetPicker
        workspaceSlug={WORKSPACE.slug}
        onChange={onChange}
        value={""}
      />,
    );

    const items = WORKSPACE.datasets.items;

    await user.click(await screen.findByTestId("combobox-button"));
    const option = await screen.queryAllByRole("option");
    expect(option.length).toBe(items.length);

    expect(container).toMatchSnapshot();
  });
});
