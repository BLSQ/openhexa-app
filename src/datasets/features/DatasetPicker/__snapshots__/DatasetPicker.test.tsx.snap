// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetPicker display all dataset 1`] = `
<div>
  <div
    class="relative"
    data-headlessui-state="open"
    data-open=""
  >
    <div
      class="form-input flex w-full items-center rounded-md border-gray-300 shadow-sm disabled:border-gray-300 focus-within:outline-none focus:ring-transparent focus-visible:border-blue-500 disabled:cursor-not-allowed  sm:text-sm border-blue-500"
    >
      <div
        class="mr-1 flex flex-1 items-center truncate"
      >
        <input
          aria-activedescendant="headlessui-combobox-option-ae7c7a26-e9c8-47d1-b9f8-3d4b8f9d0f15"
          aria-autocomplete="list"
          aria-controls="headlessui-combobox-options-d88d149f-80c3-4c6d-9b82-01013b801f93"
          aria-expanded="true"
          autocomplete="off"
          class="flex-1 placeholder-gray-600 placeholder-opacity-70 outline-none"
          data-focus=""
          data-headlessui-state="open focus"
          data-open=""
          data-testid="combobox-input"
          id="headlessui-combobox-input-b608a8f9-e22c-462e-95cb-e9c54e6fa91f"
          placeholder="Select dataset"
          role="combobox"
          type="text"
        />
      </div>
      <button
        aria-controls="headlessui-combobox-options-d88d149f-80c3-4c6d-9b82-01013b801f93"
        aria-expanded="true"
        aria-haspopup="listbox"
        data-active=""
        data-headlessui-state="open active hover"
        data-hover=""
        data-open=""
        data-testid="combobox-button"
        id="headlessui-combobox-button-963e9095-8e5e-490e-8bf2-2eb1cceb1000"
        tabindex="-1"
        type="button"
      >
        <div
          class="ml-1 flex items-center gap-0.5 rounded-r-md text-gray-400 focus:outline-none"
        >
          <svg
            class="animate-spin h-5 w-5"
            data-testid="spinner"
            fill="none"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              class="opacity-25"
              cx="12"
              cy="12"
              r="10"
              stroke="currentColor"
              stroke-width="4"
            />
            <path
              class="opacity-75"
              d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
              fill="currentColor"
            />
          </svg>
        </div>
      </button>
    </div>
    <div
      aria-labelledby="headlessui-combobox-button-963e9095-8e5e-490e-8bf2-2eb1cceb1000"
      class="max-h-60 z-10 w-full rounded-md bg-white py-1 text-base shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none sm:text-sm flex flex-col"
      data-headlessui-state="open"
      data-open=""
      data-popper-escaped="true"
      data-popper-placement="bottom-start"
      data-popper-reference-hidden="true"
      data-testid="combobox-options"
      id="headlessui-combobox-options-d88d149f-80c3-4c6d-9b82-01013b801f93"
      role="listbox"
      style="position: absolute; left: 0px; top: 0px; --input-width: 0px; --button-width: 0px; transform: translate(0px, 4px);"
    >
      <div
        class="overflow-auto flex-1"
        role="none"
      >
        <div
          class="relative cursor-default select-none px-2 py-2 bg-blue-500 text-white"
          data-active=""
          data-focus=""
          data-headlessui-state="active focus"
          id="headlessui-combobox-option-ae7c7a26-e9c8-47d1-b9f8-3d4b8f9d0f15"
          role="option"
          tabindex="-1"
        >
          <div
            class="group flex w-full items-center"
          >
            <span
              class="flex items-center pr-4 invisible text-white"
            >
              <svg
                aria-hidden="true"
                class="h-5 w-5"
                data-slot="icon"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m4.5 12.75 6 6 9-13.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
            <span
              class="flex-1 truncate"
            >
              Dataset 1
            </span>
          </div>
        </div>
        <div
          class="relative cursor-default select-none px-2 py-2 text-gray-900"
          data-headlessui-state=""
          id="headlessui-combobox-option-0dd74c72-ee7d-46a6-a83f-4120b9c8cbc6"
          role="option"
          tabindex="-1"
        >
          <div
            class="group flex w-full items-center"
          >
            <span
              class="flex items-center pr-4 invisible text-gray-900"
            >
              <svg
                aria-hidden="true"
                class="h-5 w-5"
                data-slot="icon"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m4.5 12.75 6 6 9-13.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
            <span
              class="flex-1 truncate"
            >
              Dataset 2
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
